## <a name ="started"></a>Getting started

### <a name="docker"></a>Use docker

```sh
git clone https://github.com/qinqian/gaftools
cd gaftools && docker build -t gaftools .
gaf=input_gaf
normalgaf=normal_gaf

# tumor-only or normal-only mode
docker run -i -v $(pwd):$(pwd) gaftools gaftools getindel-cython -c 4 -m 30 -l 100 --input $(pwd)/$gaf -r 3 -p $(pwd)/${gaf/.gaf/}_mapq30_mlen100_cnt3

# tumor-normal pair mode
docker run -i -v $(pwd):$(pwd) gaftools gaftools getindel -c 4 -m 30 -l 100 --input $(pwd)/$gaf --normal $normalgaf -r 3 -p $(pwd)/${gaf/.gaf/}_mapq30_mlen100_cnt3
```

### <a name="ubuntu"></a>Use Ubuntu

```sh
sudo apt install build-essential
mamba create -n gaftools python==3.10 poetry cython==3.0.7
conda activate gaftools
git clone https://github.com/qinqian/gaftools
cd gaftools && make
gaf=input_gaf
normalgaf=normal_gaf
# tumor-only or normal-only mode
gaftools getindel-cython -m 30 -l 100 --input $gaf -r 3 -p output_prefix

# tumor-normal pair mode
docker run -i -v $(pwd):$(pwd) gaftools gaftools getindel -c 4 -m 30 -l 100 --input $(pwd)/$gaf --normal $normalgaf -r 3 -p $(pwd)/${gaf/.gaf/}_mapq30_mlen100_cnt3
```

## <a name="intro"></a>Introduction

`gaftools` is a pangenome tool that predicts germline or somatic structural variation from the long read whole genome sequencing data for tumor-normal pairs of samples, tumor-only or normal-only sample. The experiment could be either normal-only, tumor-only or tumor-normal paired sequencing.

## Table of Contents

<!--<img aligh="right" width="278" src="doc/example1.png">-->

- [Getting Started](#started)
  - [Install with docker](#docker)
  - [Install with ubuntu](#ubuntu)
- [Introduction](#intro)
- [Data preprocessing](#process)
- [Usage](#usage)


## <a name="process"></a>Data preprocessing

### Minigraph

### Minimap2

## <a name="usage"></a>Usage

```sh
-m mapping quality
-l large indel size threshold
--input gaf_or_paf input file path, generated by minigraph or minimap2
-p bed_output, output bed gzip file for all the merged indel after threshold
```
